tools/README.md

```diff
@@ -0,0 +1,391 @@
+        extra_pip_args = ["--require-hashes"],
+    )
+```
+
+Lets add an empty `tools/sometools/requirements.txt`:
+
+```console
+
+$ mkdir tools/sometools
+$ touch tools/sometools/requirements.txt
+
+```
+
+We can now use `sometools_pip3` in the `BUILD` file for our Python tool, although we will need
+some actual requirements for it to be useful.
+
+In order to ensure that this `requirements.txt` stays up-to-date we will also need to add an entry
+in `.github/dependabot.yml`.
+
+This example requires the following entry:
+
+```yaml
+  - package-ecosystem: "pip"
+    directory: "/tools/sometools"
+    schedule:
+      interval: "daily"
+```
+
+
+### Add Python requirements
+
+For the purpose of this example, `mytool.py` will have dependencies on the `requests` and `pyyaml`
+libraries.
+
+Check on pypi for the most recent versions.
+
+At the time of writing these were `2.25.1` and `5.4.1` for `requests` and `pyyaml` respectively.
+
+Add the pinned dependencies to the `tools/sometools/requirements.txt` file:
+
+```console
+
+$ echo pyyaml==5.4.1 >> tools/sometools/requirements.txt
+$ echo requests==2.25.1 >> tools/sometools/requirements.txt
+
+```
+
+So far we have not set the hashes for the requirements.
+
+The easiest way to add the necessary hashes and dependencies is to use `pip-compile` from `pip-tools`.
+
+This will pin all dependencies *of these libraries* too.
+
+Run the following to update the `requirements.txt`:
+
+```console
+
+$ pip install pip-tools
+$ pip-compile --generate-hashes tools/sometools/requirements.txt
+
+```
+
+### Add the Python tool
+
+For the purpose of this example we will add a trivial tool that dumps information
+about a Python package as `yaml`
+
+Create a file `tools/sometools/mytool.py` with the following content:
+
+
+```python
+#!/usr/bin/env python3
+
+import sys
+
+import requests
+import yaml
+
+
+def main(*args) -> int:
+    sys.stdout.write(
+        yaml.dump(
+            requests.get(f"https://pypi.python.org/pypi/{args[0]}/json").json()))
+    return 0
+
+
+if __name__ == "__main__":
+    sys.exit(main(*sys.argv[1:]))
+```
+
+### Create the `BUILD` file for `sometools`
+
+If you are adding a tool to an existing toolset you may be able skip this step -
+just make sure that `load` lines are present.
+
+Add the following content to the file `tools/sometools/BUILD`
+
+```starlark
+load("//tools/base:envoy_python.bzl", "envoy_py_binary")
+load("@sometools_pip3//:requirements.bzl", "requirement")
+
+licenses(["notice"])  # Apache 2
+```
+
+Note the loading of `requirement` from `@sometools_pip3`, and the loading of `envoy_py_binary`.
+
+We will use these in the next section.
+
+### Create the bazel target for the tool
+
+Add `mytool.py` as an `envoy_py_binary` to the `tools/sometools/BUILD` file.
+
+This will make the `mytool.py` file runnable as a `bazel` target, and will
+add a test runner to ensure the file is tested.
+
+```starlark
+envoy_py_binary(
+    name = "tools.sometools.mytool",
```

Comment:
the build rule is here unless i misunderstand

Reply:
Sorry, I was confused, I didn't realize it was auto-generating the test target while reading.